File: test/p2p.py
'''
由于时间限制，该文件目前仅支持地面站服务（锥形NAT）
Author：江榕煜
e-mail：2373180028@qq.com
'''

import socket


class P2P:
    """
    内部对象
    """
    __TCPcommand = {
        "P2P_begin" : 1,
        "P2P_OK" : 2,
        "P2P_ERROR" : 3
    }

    def __init__(self,AppID = 2020,server_IPv4 = "server.singularity-blog.top",server_TCPport = 8000,server_UDPport = 8001,TCP_port = 8000,UDP_port = 8001):
        #存储基本属性
        self.AppID = AppID
        self.server_IPv4 = server_IPv4
        self.server_UDPport = server_UDPport
        self.server_TCPport = server_TCPport
        self.local_UDPport = UDP_port
        self.local_TCPport = TCP_port

        # print("AppID is:"+str(self.AppID)+",IP of server is :"+self.server_IPv4)
        print("AppID is:%d,IP of server is :%s"%(self.AppID,self.server_IPv4))

        #创建套接字
        self.socket_UDP = socket.socket(socket.AF_INET,socket.SOCK_DGRAM)
        self.socket_TCP = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
        #设置本地套接字端口
        self.socket_UDP.bind((str(socket.INADDR_ANY),UDP_port))
        self.socket_TCP.bind((str(socket.INADDR_ANY),TCP_port))
        print("本地UDP服务启动，端口号：%d"%UDP_port)
        print("本地TCP服务启动，端口号：%d"%TCP_port)
        #连接服务器TCP
        print("开始和服务器握手")
        self.socket_TCP.connect((server_IPv4,server_TCPport))
        print("服务器握手成功（TCP）")
        #等待穿透指示（TCP）
        print("等待服务器发送穿透开始指令")
        recvData = self.socket_TCP.recv(4)
        if(recvData != self.__TCPcommand["P2P_begin"]):
            print("SERVER CMD ERROR! when 穿透指示")
            return False
        #-----------开始穿透-------------
        print("开始内网穿透...")
        #发送（注册）口令（UDP）
        self.socket_UDP.sendto(     AppID.to_bytes(length = 4,
                                                                            byteorder = 'little',
                                                                            signed = True),
                                                                            (server_IPv4,server_UDPport))
        print("口令注册")
        #等待接收作业端地址（TCP）
        recvData = self.socket_TCP.recv(16)
        self.terminal_IPv4 = recvData.decode('utf-8')
        print("从服务器接收到作业端公网IP：%s"%self.terminal_IPv4)
        #接收作业端UDP端口（TCP）
        recvData = self.socket_TCP.recv(4)
        self.terminal_UDPport = int.from_bytes(recvData,byteorder='little',signed=True)
        print("从服务器接收到作业端UDP通信端口：%d"%self.terminal_UDPport)
        #本地穿透
        self.socket_UDP.sendto(      AppID.to_bytes(length = 4,
                                                                            byteorder = 'little',
                                                                            signed = True),
                                                                            (self.terminal_IPv4,self.terminal_UDPport))
        #接收作业端的握手消息
        recvData,recvAddr = self.socket_UDP.recvfrom(4)
        recvData = int.from_bytes(recvData,byteorder='little',signed=True)
        print("接收作业端UDP握手消息：%d"%recvData)
        if(recvData != self.AppID):
            print("作业端传来的ID ERROR!")
            return False
        print("穿透成功")
        #反馈成功信息给服务器
        self.socket_TCP.send(self.__TCPcommand["P2P_OK"].to_bytes(
                                                                                                                                length = 4,
                                                                                                                                byteorder = 'little',
                                                                                                                                signed = True))
        return True
    
    def P2P_sendData(self,data):
        """
        调用该函数发送数据，注意：底层使用UDP
        """
        self.socket_UDP.sendto( data,(self.terminal_IPv4,self.terminal_UDPport))
        return

    def P2P_sendCMD(self,data):
        """
        调用该函数发送重要数据or命令，注意：底层使用TCP
        """
        self.socket_TCP.send(data)
        return

    def P2P_recvData(self,dataSize):
        """
        调用该函数接收数据，注意：底层使用UDP
        """
        return self.socket_UDP.recvfrom(dataSize)

    def P2P_recvCMD(self,dataSize):
        """
        调用该函数接收重要数据or命令
        可以选择等待数据（阻塞）或试探性读取（非阻塞）
        """
        return self.socket_TCP.recv(dataSize)
        

File: P2P_transfer/p2p.cpp
#include "p2p.hpp"

#define __USER_DEBUG_P2P_CPP__

/*	@brief:	    通信初始化（进行TCP握手，UDP穿透）
 *	@notice:    注意在使用该函数之前需要对服务器地址进行初始化
 *	@author:    江榕煜（2020.10.3）
 *	@param:     
                p2p对象
 *	@retv:	        创建成功or失败
**/
bool P2P_Init(p2p * target)
{
    int sendRet,recvRet;
    int recvAppID;
    int addrLength = sizeof(target->addr_recv);
    switch (target->localType)
    {
    case TERMINAL:
        target->socket_UDP = socket(AF_INET,SOCK_DGRAM,0);
        target->socket_TCP = socket(AF_INET,SOCK_STREAM,0);
        if(target->socket_UDP == -1 || target->socket_TCP == -1)
        {
            return false;
        }

        /*****************调试辅助打印(START)********************/
        #ifdef __USER_DEBUG_P2P_CPP__
        std::cout<<"套接字创建成功!"<<std::endl;
        #endif
        /*****************调试辅助打印(END)********************/

        //初始化UDP套接字，绑定端口
        target->addr_send.sin_family = AF_INET;     //IPv4地址
        target->addr_send.sin_addr.s_addr = htonl(INADDR_ANY);                  //任意对象
        target->addr_send.sin_port = htons(target->port_server_UDP);  //本地开放端口
        if( bind(   target->socket_UDP,
                            (struct sockaddr *)&(target->addr_send),
                            sizeof(target->addr_send)   ) 
            < 0)
        {
            return false;
        }
        /*****************调试辅助打印(START)********************/
        #ifdef __USER_DEBUG_P2P_CPP__
        std::cout<<"UDP套接字绑定端口!"<<std::endl;
        #endif
        /*****************调试辅助打印(END)********************/

        //TCP套接字绑定端口
        target->addr_send.sin_addr.s_addr = htonl(INADDR_ANY);
        target->addr_send.sin_port = htons(target->port_server_TCP);
        if(bind(  target->socket_TCP,
                            (struct sockaddr *)&(target->addr_send),
                            sizeof(target->addr_send)   )
                < 0)
        {
            return false;
        }

        /*****************调试辅助打印(START)********************/
        #ifdef __USER_DEBUG_P2P_CPP__
        std::cout<<"TCP套接字端口绑定!"<<std::endl;
        #endif
        /*****************调试辅助打印(END)********************/


        /*****************调试辅助打印(START)********************/
        #ifdef __USER_DEBUG_P2P_CPP__
        std::cout<<"等待和TCP服务器握手"<<std::endl;
        #endif
        /*****************调试辅助打印(END)********************/

        //TCP套接字与服务器握手
        target->addr_send.sin_family = AF_INET;     //IPv4地址
        target->addr_send.sin_addr.s_addr = inet_addr(target->IPv4_server_TCP); //服务器地址
        target->addr_send.sin_port = htons(target->port_server_TCP);  //服务器端口
        if( connect(    target->socket_TCP,           //与服务器TCP握手
                        (struct sockaddr *)&(target->addr_send),
                        sizeof(target->addr_send)   ) 
            == -1)
        {
                /*****************调试辅助打印(START)********************/
                #ifdef __USER_DEBUG_P2P_CPP__
                std::cout<<"TCP服务器握手 ERROR!"<<std::endl;
                #endif
                /*****************调试辅助打印(END)********************/

            return false;
        }

        /*****************调试辅助打印(START)********************/
        #ifdef __USER_DEBUG_P2P_CPP__
        std::cout<<"TCP服务器握手 成功!"<<std::endl;
        #endif
        /*****************调试辅助打印(END)********************/

        //注册设备口令(UDP)
        target->addr_send.sin_addr.s_addr = inet_addr(target->IPv4_server_UDP);    //目标服务器IP
        target->addr_send.sin_port    = htons(target->port_server_UDP);           //端口设置导入
        sendRet = sendto(target->socket_UDP,        //UDP套接字
                        &target->APP_ID,             //口令（app号）
                        sizeof(target->APP_ID),     //口令大小
                        0,  //flag
                        (struct sockaddr *)&(target->addr_send),    //发送服务器
                        sizeof(target->addr_send));
        
        /*****************调试辅助打印(START)********************/
        #ifdef __USER_DEBUG_P2P_CPP__
        std::cout<<"terminal UDP send AppID:"<< target->APP_ID << " success!"<<std::endl;
        std::cout<<"等待服务器UDP响应"<<std::endl;
        #endif
        /*****************调试辅助打印(END)********************/

        recvRet = recvfrom(target->socket_UDP,&recvAppID,sizeof(recvAppID),0,
                            (sockaddr *)&(target->addr_recv),(socklen_t *)&(addrLength));
        memcpy(&(target->addrCache),&(target->addr_recv.sin_addr.s_addr),4);
        std::cout<<"接受到响应，IP："<<inet_ntoa(target->addrCache)
            <<".端口："<<ntohs(target->addr_recv.sin_port)<<std::endl;

        /*****************调试辅助打印(START)********************/
        #ifdef __USER_DEBUG_P2P_CPP__
        std::cout<<"P2P穿透FINISH"<<std::endl;
        #endif
        /*****************调试辅助打印(END)********************/

        break;
    
    default:
        break;
    }
}

/*	@brief:	        发送数据给对方
 *	@notice:        用的是UDP
 *	@author:    江榕煜（2020.10.3）
 *	@param:     
                target（p2p对象）
                data（指针）发送数据的头指针
                dataLength（size_t）数据长度
 *	@retv:	        发送成功or失败
**/
bool P2P_sendData(p2p *target,void * data,size_t dataLength)
{
    int sendRet;
    switch (target->localType)
    {
    case TERMINAL:
        target->addr_send.sin_family = AF_INET;     //IPv4地址
        target->addr_send.sin_addr.s_addr = inet_addr(target->IPv4_server_UDP);    //终端发向地面站
        target->addr_send.sin_port = htons(target->port_server_UDP);  //地面站端口
        
        sendRet = sendto(target->socket_UDP,
                            data,dataLength,
                            0,
                            (sockaddr *)&(target->addr_send),
                            sizeof(target->addr_send));
        if(sendRet == -1)
            return false;
        break;

    default:
        return false;
        break;
    }

    return true;
}

/*	@brief:	        发送命令or重要数据给对方
 *	@notice:        用的是TCP
 *	@author:    江榕煜（2020.10.3）
 *	@param:     
                target（p2p对象）
                data（指针）发送数据的头指针
                dataLength（size_t）数据长度
 *	@retv:	        发送成功or失败
**/
bool P2P_sendCMD(p2p *target,void * data,size_t dataLength)
{
    int sendRet;
    int socketNow;
    switch (target->localType)
    {
    case TERMINAL:
        sendRet = send(target->socket_TCP,
                        data,dataLength,0);
        if(sendRet == -1)
            return false;
        break;

    default:
        return false;
        break;
    }

    return true;

}

/*	@brief:	        接收数据
 *	@notice:        用的是UDP
 *	@author:    江榕煜（2020.10.3）
 *	@param:     
                target（p2p对象）
                data（指针）数据接收池的头指针
                dataLength（size_t）接收池长度
 *	@retv:	        发送成功or失败
**/
long P2P_recvData(p2p *target,void * data,size_t dataLength)
{
    int addrLength = sizeof(target->addr_recv);
    int recvRet;
    switch (target->localType)
    {
    case TERMINAL:
        do
        {
            recvRet = recvfrom(target->socket_UDP,
                                data,dataLength,
                                0,
                                (sockaddr *)&(target->addr_recv),
                                (socklen_t *)&addrLength);
            if(recvRet == -1)
                return -1;
            if(target->addr_recv.sin_addr.s_addr != inet_addr(target->IPv4_server_UDP))
                continue;
        } while (0);
        
        
        break;

    default:
        return -1;
        break;
    }

    return recvRet;

}

/*	@brief:	        接收命令or重要数据
 *	@notice:        用的是TCP
 *	@author:    江榕煜（2020.10.3）
 *	@param:     
                target（p2p对象）
                data（指针）数据接收池的头指针
                dataLength（size_t）接收池长度
 *	@retv:	        发送成功or失败
**/
long P2P_recvCMD(p2p *target,void * data,size_t dataLength)
{
    int addrLength = sizeof(target->addr_recv);
    int recvRet;
    switch (target->localType)
    {
    case TERMINAL:
        recvRet = recv(target->socket_TCP,
                        data,dataLength,0);
        if(recvRet == -1)
            return -1;
        
        break;

    default:
        return -1;
        break;
    }

    return recvRet;
    
}

/*	@brief:	        清理P2P对象使用的网络内核栈
 *	@notice:        此函数谨慎使用！BE CAREFUL！
 *	@author:      江榕煜（2020.10.6）
 *	@param:
                target（p2p对象）
 *	@retv:	            无
**/
void P2P_stackClean(p2p*target)
{
    static char history[1024];
    static int addrLength = sizeof(target->addr_recv);
    while (recv(target->socket_TCP,history,1024,MSG_DONTWAIT) == 1024){}
    while (recvfrom(target->socket_UDP,
                        history,1024,
                        0,
                        (sockaddr *)&(target->addr_recv),
                        (socklen_t *)&addrLength)==1024){}
    return;
}
